File: new/150_evaluate_reverse_polish_notation_stack.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        # idea: stack
        # loop through the tokens
        # have a stack tracking all the operands
        # in the loop if the current token is an operand add it to the stack.
        # if the current token is and operator, pop the previous 2 elements and do the operation
        # for those 2 elements in correct order.
        stack = []
        for token in tokens:
            if token == "+":
                b = stack.pop()
                a = stack.pop()
                stack.append(a + b)
            elif token == "-":
                b = stack.pop()
                a = stack.pop()
                stack.append(a - b)
            elif token == "*":
                b = stack.pop()
                a = stack.pop()
                stack.append(a * b)
            elif token == "/":
                b = stack.pop()
                a = stack.pop()
                stack.append(int(a / b))
            else:
                stack.append(int(token))

        return stack[0]
        # time: O(n)
        # space: O(n)

File: new/cpp/150_evaluate_reverse_polish_notation_stack.cpp
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        // idea: stack
        // loop through each token in tokens, if the token is a operand, add it to our stack 
        // this stack tracks all the numbers
        // once we hit an operand in the loop, pop the top 2 numbers from the stack, and do the operation
        // in correct order. 
        stack<int> st;
        for (string& token: tokens) {
            if (token == "+") {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                st.push(a + b);
            } else if (token == "-") {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                st.push(a - b);
            } else if (token == "*") {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                st.push(a * b);
            } else if (token == "/") {
                int b = st.top();
                st.pop();
                int a = st.top();
                st.pop();
                st.push(a / b);
            } else {
                st.push(stoi(token));
            }
        }

        return st.top();
    }
    // time: O(n)
    // space: O(n)
};
